File: charactermanager/backend/src/main/java/ai/ragu/charactermanager/enumeration/RangedWeaponTypeEnum.java
package ai.ragu.charactermanager.enumeration;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonValue;

import java.util.HashMap;
import java.util.Map;

public enum RangedWeaponTypeEnum implements AttackProficiencyEnum {
    DART("dart"),
    BOW("bow");

    private final String value;
    private final static Map<String, RangedWeaponTypeEnum> CONSTANTS = new HashMap<>();

    static {
        for (RangedWeaponTypeEnum c: values()) {
            CONSTANTS.put(c.value, c);
        }
    }

    RangedWeaponTypeEnum(String value) {
        this.value = value;
    }

    @Override
    public String toString() {
        return this.value;
    }

    @JsonValue
    public String value() {
        return this.value;
    }

    @JsonCreator
    public static RangedWeaponTypeEnum fromValue(String value) {
        RangedWeaponTypeEnum constant = CONSTANTS.get(value);
        if (constant == null) {
            throw new IllegalArgumentException(value);
        } else {
            return constant;
        }
    }
}

File: charactermanager/backend/src/main/java/ai/ragu/charactermanager/enumeration/RangedWeaponTypeEnum.kt
package ai.ragu.charactermanager.enumeration

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonValue
import java.util.*

enum class RangedWeaponTypeEnum(private val value: String) : AttackProficiencyEnum {
    DART("dart"), BOW("bow");

    companion object {
        private val CONSTANTS: MutableMap<String, RangedWeaponTypeEnum> = HashMap()
        @JsonCreator
        fun fromValue(value: String?): RangedWeaponTypeEnum {
            val constant = CONSTANTS[value]
            return constant ?: throw IllegalArgumentException(value)
        }

        init {
            for (c in values()) {
                CONSTANTS[ai.ragu.charactermanager.enumeration.c.value] = ai.ragu.charactermanager.enumeration.c
            }
        }
    }

    override fun toString(): String {
        return value
    }

    @JsonValue
    fun value(): String {
        return value
    }

}